Introduce CVA to style TwigComponent

WebMamba opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Issues        | 
| License       | MIT

This PR introduces a new concept CVA (Class Variance Authority), by adding a1 twig function, to help you manage your class in your component.

Let's take an example an Alert component. In your app, an alert can have a lot of different styles one for success, one for alert, one for warning, and different sizes, with icons or not... You need something that lets you completely change the style of your component without creating a new component, and without creating too much complexity in your template.

Here is the reason came CVA.

Your Alert component can now look like this:

```twig
{% props color = 'blue', size = 'md' %}

{% set alert =  cva({
     base: 'alert rounded-lg',
     variants: {
        color: {
            blue: 'text-blue-800 bg-blue-50 dark:bg-gray-800 dark:text-blue-400',
            red: 'text-red-800 bg-red-50 dark:bg-gray-800 dark:text-red-400',
            green: 'text-green-800 bg-green-50 dark:bg-gray-800 dark:text-green-400',
            yellow: 'text-yellow-800 bg-yellow-50 dark:bg-gray-800 dark:text-yellow-400',
        },
        size: {
            sm: 'px-4 py-3 text-sm',
            md: 'px-6 py-4 text-base',
            lg: 'px-8 py-5 text-lg',
        }
    },
    compoundVariants: [{
           color: ['red'],
           size: ['lg'],
          class: 'font-semibold'
    }],
    defaultVariants: {
           rounded: 'md'
    }
}) %}

<div class="{{ cva.apply({color, size}, attribute.render('class'), 'flex p-4') }}">
    ...
</div>
```

So here you have a `cva` function that lets you define different variants of your component.

You can now use your component like this:
```twig
<twig:Alert color="red" size="md"/>

<twig:Alert color="green" size="sm"/>

<twig:Alert color="yellow" size="lg"/>

<twig:Alert color="red" size="md" class="dark:bg-gray-800"/>
```

And then you get the following result:

<img width="1269" alt="Capture d’écran 2024-01-24 à 00 52 33" src="https://github.com/symfony/ux/assets/32077734/6a5e25be-5b81-4ae7-8385-0fa5422d0396">

If you want to know more about the concept I implement here you can look at:
- CVA (js version): https://cva.style/docs
- tailwind merge: https://github.com/gehrisandro/tailwind-merge-php, https://github.com/dcastil/tailwind-merge
- this implementation by using tailwind-merge and cva is inspired a lot by: https://ui.shadcn.com/ (shadcn is the most starred library on github in 2023) 
- a really good article that explains the philosophy behind https://manupa.dev/blog/anatomy-of-shadcn-ui
- this PR works great in a LASTstack: https://symfonycasts.com/screencast/last-stack/last-stack

Tell me what you think about it! Thanks for your time! Cheers 🧡

WebMamba committed:
Introduce CVA to style TwigComponent

kbond commented:
Can you show what `twm` and `clm` stand for?

Is there resources to provide some context on `cva`, `twm`, and `clm`?

Kocal commented on src/TwigComponent/src/Twig/ComponentExtension.php:
```diff
@@ -84,6 +89,43 @@ public function finishEmbeddedComponentRender(): void
 
+    public function cva(array &$context, string $recipeName, array $recipe): void
+    {
+        $key = '__recipe__'.$recipeName;
+
+        $context[$key] = CVA::init($recipe);
```
Do you think it is possible to use the Twig filename to combine it to the context key?

I'm afraid of conflicts between _too-much generic name_ recipe name across multiple components.

**EDIT:** or this is what `needs_context` and `$context` already do?

WebMamba replied:
yes the context is only for a specific file

Kocal replied:
Interesting, I've learned something today :d

WebMamba replied:
Actually not a specific file, but every component live inside his own context

Kocal commented on src/TwigComponent/tests/Fixtures/templates/components/Alert.html.twig:
```diff
@@ -0,0 +1,22 @@
+{% props color = 'blue', size = 'md', class = '' %}
+
+{{ cva('alert', {
+    'base': 'rounded-lg',
+    'variants': {
+        'color': {
+            'blue': 'text-blue-800 bg-blue-50 dark:bg-gray-800 dark:text-blue-400',
+            'red': 'text-red-800 bg-red-50 dark:bg-gray-800 dark:text-red-400',
+            'green': 'text-green-800 bg-green-50 dark:bg-gray-800 dark:text-green-400',
+            'yellow': 'text-yellow-800 bg-yellow-50 dark:bg-gray-800 dark:text-yellow-400',
+        },
+        'size': {
+            'sm': 'px-4 py-3 text-sm',
+            'md': 'px-6 py-4 text-base',
+            'lg': 'px-8 py-5 text-lg',
+        }
+    }
+}) }}
```
```suggestion
{% cva('alert', {
    'base': 'rounded-lg',
    'variants': {
        'color': {
            'blue': 'text-blue-800 bg-blue-50 dark:bg-gray-800 dark:text-blue-400',
            'red': 'text-red-800 bg-red-50 dark:bg-gray-800 dark:text-red-400',
            'green': 'text-green-800 bg-green-50 dark:bg-gray-800 dark:text-green-400',
            'yellow': 'text-yellow-800 bg-yellow-50 dark:bg-gray-800 dark:text-yellow-400',
        },
        'size': {
            'sm': 'px-4 py-3 text-sm',
            'md': 'px-6 py-4 text-base',
            'lg': 'px-8 py-5 text-lg',
        }
    }
}) %}
```

WebMamba replied:
Much harder to do but yes maybe I gonna be have to do it that way

Kocal replied:
Why is it much harder to do? (it's not a trap, that's a real question :d)

smnandre replied:
If you need any help @WebMamba poke me, but i'd also prefer we keep Twig standard concerning tags/functions, ... and if your CSV does not echo anything i would be better to use a tag.

Kocal commented on src/TwigComponent/tests/Fixtures/templates/components/Alert.html.twig:
```diff
@@ -0,0 +1,22 @@
+    }
+}) }}
+
+<div class="{{ twm('alert', {color, size}, class) }}">
```
Don't you think `twm` (for **T**ail**w**ind **M**erge) and `clm` (for **Cl**asses **M**erge) should only accept classes as parameter, and not the variant name/values? (this is what https://github.com/dcastil/tailwind-merge#tailwind-merge does)

IMO those functions have too much responsability.

```suggestion
{% set styles_alert = cva({
    'base': 'rounded-lg',
    'variants': {
        'color': {
            'blue': 'text-blue-800 bg-blue-50 dark:bg-gray-800 dark:text-blue-400',
            'red': 'text-red-800 bg-red-50 dark:bg-gray-800 dark:text-red-400',
            'green': 'text-green-800 bg-green-50 dark:bg-gray-800 dark:text-green-400',
            'yellow': 'text-yellow-800 bg-yellow-50 dark:bg-gray-800 dark:text-yellow-400',
        },
        'size': {
            'sm': 'px-4 py-3 text-sm',
            'md': 'px-6 py-4 text-base',
            'lg': 'px-8 py-5 text-lg',
        }
    }
}) %}


<div class="{{ twm(cva(styles_alert, { color, size }), class, class2, class3) }}">
```

Here, `cva({})` returns an instance of `CVA`, and if you pass it back to `cva` (`cva(styles_alert, { color, size })`, then it will render your CSS classes.

WDYT?

WebMamba replied:
I did that at first but Ryan didn't like it 😅

CMH-Benny replied:
If using strings does it become an option to pass multiple cva names?
Like:
```
<div class="{{ twm('alert other_cva', {color, size, variant_from_other_cva}, class)">
```
Or is this too advanced for now?

CMH-Benny replied:
I thought about the main question here and after some thinking I would say I agree, in this implementation twm() is only usable together with cva, because it always expect a cva as first parameter, which not only combines two things that are independant from each other, it also means anyone who just wants to use some tw classes in a much simpler component can't use this twm to merge tw classes, but has to use cva with base styles for that.

I wonder what Ryan didn't like about separating those two steps 🤔 
After some thinking, what about something like this:

```twig
{% set other_classes = 'm-4 grid grid-rows-[1fr]' %}
{% set alert = cva({
    'base': 'rounded-lg',
    'variants': {
        'color': {
            'blue': 'text-blue-800 bg-blue-50 dark:bg-gray-800 dark:text-blue-400',
            'red': 'text-red-800 bg-red-50 dark:bg-gray-800 dark:text-red-400',
            'green': 'text-green-800 bg-green-50 dark:bg-gray-800 dark:text-green-400',
            'yellow': 'text-yellow-800 bg-yellow-50 dark:bg-gray-800 dark:text-yellow-400',
        },
        'size': {
            'sm': 'px-4 py-3 text-sm',
            'md': 'px-6 py-4 text-base',
            'lg': 'px-8 py-5 text-lg',
        }
    }
}) %}

{% set alert_classes = cvapply(alert, { color, size }) %}
<div class="{{ twm(altert_classes, other_classes, more_tw_classes) }}">
```

It allows for more flexible usage, one could also use multiple cva, get the classes based on params and then throw all of them into twm, which allows some kind of composing, WDYT?

Kocal commented on src/TwigComponent/src/Twig/ComponentExtension.php:
```diff
@@ -84,6 +89,43 @@ public function finishEmbeddedComponentRender(): void
+            return '';
+        }
+
+        $tw = TailwindMerge::instance();
```
`TailwindMerge` should be configurable, because some projects can use a prefix (e.g.: `tw-`) for easily transitioning to Tailwind, and I don't think this will work out-of-the-box.

WebMamba replied:
Yes sure this a first draft, to show the ideas, but yeah I have in mind to add configuration and caching

Kocal reviewed (commented):
Good :D

norkunas commented:
I'd probably better have something like https://styled-components.com/ or https://emotion.sh/docs/introduction because now it looks like that ux is going all tailwind :stuck_out_tongue:

WebMamba commented:
@kbond description updated! 😁

> I'd probably better have something like https://styled-components.com/ or https://emotion.sh/docs/introduction because now it looks like that ux is going all tailwind 😛

Yes, go LASTstack! No, you can actually use all of this without tailwind by using `clm`instead of `twm`but yeah I think you have to architecture your style in a class-utility way. I don't believe in styled-components and emotion because you write css in to string so you don't have proper autocomplete, and  it requires another heavy build step.

norkunas commented:
> @kbond description updated! 😁
> 
> > I'd probably better have something like https://styled-components.com/ or https://emotion.sh/docs/introduction because now it looks like that ux is going all tailwind 😛
> 
> Yes, go LASTstack! No, you can actually use all of this without tailwind by using `clm`instead of `twm`but yeah I think you have to architecture your style in a class-utility way. I don't believe in styled-components and emotion because you write css in to string so you don't have proper autocomplete, and  it requires another heavy build step.

Objects could be allowed instead of string :) no need for build step,as css could be inlined into head

kbond commented:
What about the following syntax:

```twig
{% props color = 'blue', size = 'md' %}

<div{{ attributes.defaults({
    class: cva('alert', {
        'base': 'rounded-lg',
        'variants': {
            'color': {
                'blue': 'text-blue-800 bg-blue-50 dark:bg-gray-800 dark:text-blue-400',
                'red': 'text-red-800 bg-red-50 dark:bg-gray-800 dark:text-red-400',
                'green': 'text-green-800 bg-green-50 dark:bg-gray-800 dark:text-green-400',
                'yellow': 'text-yellow-800 bg-yellow-50 dark:bg-gray-800 dark:text-yellow-400',
            },
            'size': {
                'sm': 'px-4 py-3 text-sm',
                'md': 'px-6 py-4 text-base',
                'lg': 'px-8 py-5 text-lg',
            }
        }
    }).apply(color, size),
})>
    ...
</div>
```

clm/twm functions wouldn't exist but be applied automatically.

WebMamba commented:
I am not sure about this syntax, because inside your div you can have other elements to tag with recipes and things can get really big and confusing. If I made two functions it's because I want the style to be separate from html, to make everything more readable.

WebMamba commented:
> Objects could be allowed instead of string :) no need for build step,as css could be inlined into head

Actually, the idea is not to make the perfect PR to deal with your component style, but to give utilities to manage your class. Maybe in another PR we can think about CSS in components, but this is not the topic of this PR. And actually don't believe in CSS in components but you can still convince me 😁

norkunas commented:
> Actually, the idea is not to make the perfect PR to deal with your component style, but to give utilities to manage your class. Maybe in another PR we can think about CSS in components, but this is not the topic of this PR.

Yes, sorry, just wanted to bring this in :grin: 

> And actually don't believe in CSS in components but you can still convince me 😁

Why not ? :) if you don't use css frameworks, then it would be nice to have "non detached" css from components :) and that could work for every element not only the components :)

WebMamba commented:
I am playing with the ideas from @CMH-Benny  https://github.com/symfony/ux/pull/1416#discussion_r1465072527 and @kbond and @Kocal 
What do you think of that:
```twig
{% props color = 'blue', size = 'md', class = '' %}

{% set alert =  cva('alert', {
    'base': 'rounded-lg',
    'variants': {
        'color': {
            'blue': 'text-blue-800 bg-blue-50 dark:bg-gray-800 dark:text-blue-400',
            'red': 'text-red-800 bg-red-50 dark:bg-gray-800 dark:text-red-400',
            'green': 'text-green-800 bg-green-50 dark:bg-gray-800 dark:text-green-400',
            'yellow': 'text-yellow-800 bg-yellow-50 dark:bg-gray-800 dark:text-yellow-400',
        },
        'size': {
            'sm': 'px-4 py-3 text-sm',
            'md': 'px-6 py-4 text-base',
            'lg': 'px-8 py-5 text-lg',
        }
    }
}) }} %}

<div class="{{ alert.apply({color, size})|tw_merge }}">
    ...
</div>
```

Kocal commented:
That's really great, I prefer this solution!

Also, do you need to pass `'alert'` as `cva()` 1st parameter?

CMH-Benny commented:
> I am playing with the ideas from @CMH-Benny [#1416 (comment)](https://github.com/symfony/ux/pull/1416#discussion_r1465072527) and @kbond and @Kocal What do you think of that:
> 
> ```
> {% props color = 'blue', size = 'md', class = '' %}
> 
> {% set alert =  cva('alert', {
>     'base': 'rounded-lg',
>     'variants': {
>         'color': {
>             'blue': 'text-blue-800 bg-blue-50 dark:bg-gray-800 dark:text-blue-400',
>             'red': 'text-red-800 bg-red-50 dark:bg-gray-800 dark:text-red-400',
>             'green': 'text-green-800 bg-green-50 dark:bg-gray-800 dark:text-green-400',
>             'yellow': 'text-yellow-800 bg-yellow-50 dark:bg-gray-800 dark:text-yellow-400',
>         },
>         'size': {
>             'sm': 'px-4 py-3 text-sm',
>             'md': 'px-6 py-4 text-base',
>             'lg': 'px-8 py-5 text-lg',
>         }
>     }
> }) }} %}
> 
> <div class="{{ alert.apply({color, size})|tw_merge }}">
>     ...
> </div>
> ```
I like it!
You know what, if tw_merge becomes a filter instead of a function, you could even leave it out, so people that don't use tailwind don't have that dependency without a choice but if they tailwind, they could install https://github.com/tales-from-a-dev/twig-tailwind-extra It makes it harder to setup, but it would be more modual. Just a thought :) 

What do you think?

I have another question, if i do:
```twig
{{ alert.apply({color, size})|tw_merge }}
```
it will automatically pass it as key value pairs and would `alert.apply({color})` also work if I don't want the size classes and `alert.apply({size})`  if I only want the size classes?

I am not deep into the twig syntax, to me {} is an assoc array notation and I am not aware if it's similar to JS where you can put a variable `{ x, y, z }` and it will automatically become `{ x: x, y: y, z: z }` - So is:
`alert.apply({color}) === alert.apply({color: color})` ?

Kocal commented:
> I am not deep into the twig syntax, to me {} is an assoc array notation and I am not aware if it's similar to JS where you can put a variable `{ x, y, z }` and it will automatically become `{ x: x, y: y, z: z }` - So is: `alert.apply({color}) === alert.apply({color: color})` ?

Yus, like JavaScript `{ color }` expands to `{ color: color }`.

WebMamba commented:
> Also, do you need to pass 'alert' as cva() 1st parameter?

No, it's typo

> t makes it harder to setup, but it would be more modual. Just a thought :) What do you think?

I am agree. 😁

I gonna try do to a new proposal this week-end

weaverryan commented:
Hi!

Great iterating here! If I understand correctly, the currently-agreed-upon version that @WebMamba is going to work on is this:

```twig
{% props color = 'blue', size = 'md' %}

{% set alert =  cva('alert', {
    base: 'rounded-lg',
    variants: {
        color: {
            blue: 'text-blue-800 bg-blue-50 dark:bg-gray-800 dark:text-blue-400',
            red: 'text-red-800 bg-red-50 dark:bg-gray-800 dark:text-red-400',
            green: 'text-green-800 bg-green-50 dark:bg-gray-800 dark:text-green-400',
            yellow: 'text-yellow-800 bg-yellow-50 dark:bg-gray-800 dark:text-yellow-400',
        },
        size: {
            sm: 'px-4 py-3 text-sm',
            md: 'px-6 py-4 text-base',
            lg: 'px-8 py-5 text-lg',
        }
    }
}) }} %}

<div class="{{ alert.apply({color, size})|tw_merge }}">
    ...
</div>
```

(note, I removed all of the `'` in the keys - they're not needed, and this looks better 😉 ).

If so, yes, I like this! But 2 things remain:

A) How can this work with `attributes`? Like this:

```
<div {{ attributes.defaults({
    class: alert.apply({color, size})|tw_merge
}) }}>
```

B) And related, instead of creating a new `tw_merge` function, can we just tell people to use https://github.com/tales-from-a-dev/twig-tailwind-extra? AND, if that package is installed, could we automatically "merge" the `class` attribute on all `ComponentAttributes`? I think this may be tricky from an implementation standpoint, but if it's the best API, we should make it happen.

kbond commented:
> And related, instead of creating a new tw_merge function, can we just tell people to use https://github.com/tales-from-a-dev/twig-tailwind-extra?

+1 to this

WebMamba commented:
Hey! Thanks everyone for all of your ideas! So here is the details of what I did in my last commit:
- no more dependencie on https://github.com/tales-from-a-dev/twig-tailwind-extra
- remove twm function
- remove clm function
- use cva in a setter
Here is how to use CVA now (from the tests implemented in this PR):
```twig
{% props color = 'blue', size = 'md', class = '' %}

{% set alert = cva({
    base: 'alert',
    variants: {
        color: {
            blue: 'alert-blue',
            red: 'alert-red',
            green: 'alert-green',
            yellow: 'alert-yellow',
        },
        size: {
            sm: 'alert-sm',
            md: 'alert-md',
            lg: 'alert-lg',
        }
    },
    compounds: [
        {
            color: ['red'],
            size: ['lg'],
            class: 'font-semibold'
        }
    ]
}) %}

<div {{ attributes.defaults({
    class: alert.apply({color, size}, class)
}) }}>
    ...
</div>
```
Then you can use your component like so:
```twig
<twig:Alert2 color='red' size='lg' class='dark:bg-gray-600'/>
```
And you will get the following result:
```html
<div  class="alert alert-red alert-lg font-semibold dark:bg-gray-600">
    ...
</div>
```

I didn't introduce compounds in my description but the idea is if the condition in the compounds are fullfield then the classes are apply.

CMH-Benny commented:
Looks pretty good to me 👍 

It says it needs review - I am pretty new to follow PRs like this 😅  
Is there anything a rando like me can do to help to move forward or will this be reviewed as soon as somebody from the maintainers finds the time?

kbond commented on src/TwigComponent/tests/Fixtures/templates/tailwind_merge.html.twig:
```diff
@@ -0,0 +1 @@
+<twig:Alert color='red' size='lg' class='dark:bg-gray-600'/>
```
Is this template and `Alert.html.twig` used anymore?

CMH-Benny replied:
CVA = Class Variance Authority
It's an acronym from JS world: https://cva.style/docs

The idea is to have the same tool available in twig, to write components that could be similar to the (react) components from [shadcn](https://ui.shadcn.com/docs/components/alert) (Open "Manual" in Installation section and expand the code, there you can see how they use CVA in JS context, it's the exact same idea that is provided here)

smnandre replied:
I think this file should be removed for now :)

weaverryan replied:
Ping @WebMamba is this being used anymore or can it be removed?

kbond reviewed (commented):
I like this! Few things I'd like to discuss:
1. `CVA`: is this a commonly known acronym? It feels a bit arbitrary to me anyway. I'd prefer it be expanded out.
2. Since this isn't really twig-component specific (I don't think), should we at least try and get this into `twig/html-extra`? I know that is easier said than done.

smnandre commented:
I need to insist on that point, but if `cva` does not "output" something, it should be either a tag or in this case... a filter

> Most of the time though, a tag is not needed:
    * If your tag generates some output, use a function instead.
    * If your tag modifies some content and returns it, use a filter instead.
    For instance, if you want to create a tag that converts a Markdown formatted text to HTML, create a markdown filter instead: (...)

https://twig.symfony.com/doc/3.x/advanced.html#tags

So it should be

```twig

{% set alert = {
    base: 'alert',
    variants: {
        color: {
            blue: 'alert-blue',
            red: 'alert-red',
            green: 'alert-green',
            yellow: 'alert-yellow',
        },
        size: {
            sm: 'alert-sm',
            md: 'alert-md',
            lg: 'alert-lg',
        }
    },
    compounds: [
        {
            color: ['red'],
            size: ['lg'],
            class: 'font-semibold'
        }
    ]
}|cva %}

```

:)

CMH-Benny commented:
Somehow I can't reply directly, so I need to send a new comment 🙈 

> I need to insist on that point, but if `cva` does not "output" something, it should be either a tag or in this case... a filter
> 
> > Most of the time though, a tag is not needed:
> > * If your tag generates some output, use a function instead.
> > * If your tag modifies some content and returns it, use a filter instead.
> > For instance, if you want to create a tag that converts a Markdown formatted text to HTML, create a markdown filter instead: (...)
> 
> https://twig.symfony.com/doc/3.x/advanced.html#tags
> 
> So it should be
> 
> ```
> {% set alert = {
>     base: 'alert',
>     variants: {
>         color: {
>             blue: 'alert-blue',
>             red: 'alert-red',
>             green: 'alert-green',
>             yellow: 'alert-yellow',
>         },
>         size: {
>             sm: 'alert-sm',
>             md: 'alert-md',
>             lg: 'alert-lg',
>         }
>     },
>     compounds: [
>         {
>             color: ['red'],
>             size: ['lg'],
>             class: 'font-semibold'
>         }
>     ]
> }|cva %}
> ```
> 
> :)

But it outputs something now, it's a CVA object you call apply on to output the fitting classes based on given properties, so a function now makes sense? Otherwise you would have to do:

```twig
{% props color='blue', size = 'md', class = '' %}
{% set alert = {
     base: 'alert',
     variants: {
         color: {
             blue: 'alert-blue',
             red: 'alert-red',
             green: 'alert-green',
             yellow: 'alert-yellow',
         },
         size: {
             sm: 'alert-sm',
             md: 'alert-md',
             lg: 'alert-lg',
         }
     },
     compounds: [
         {
             color: ['red'],
             size: ['lg'],
             class: 'font-semibold'
         }
     ]
}|cva({color, size}, class) %}
```

And I am not sure if that works, "alert" becomes a loose object and how can the CVA now parse it and apply the correct variants? I think the provided solution looks good as it is, or do I miss anything? @smnandre

smnandre commented:
Alert is an CVA instance in both cases 

```
{# cva function #}
{% set alert = cva({foo: bar}) %}

{# cva filter #}
{% set alert = {foo: bar}|cva %}
```

Those two syntaxes happen to give the same result. But the implementation is not the same, the optimisations neither, and ... one does not follow Twig (or Symfony) conventions.

CMH-Benny commented:
Oh, now I get it - I learned something new again!

So by "output" you don't mean it doesn't return anything, it means it doesn't directly add markup/text to the final output

So if it's a function it needs to be used in `{{ cva(...) }}`
But since it's only providing an object that needs to get applied() to actually create output, it should become a filter!
So the full example would look like this:

```twig
{% props color='blue', size = 'md', class = '' %}
{% set alert = {
     base: 'alert',
     variants: {
         color: {
             blue: 'alert-blue',
             red: 'alert-red',
             green: 'alert-green',
             yellow: 'alert-yellow',
         },
         size: {
             sm: 'alert-sm',
             md: 'alert-md',
             lg: 'alert-lg',
         }
     },
     compounds: [
         {
             color: ['red'],
             size: ['lg'],
             class: 'font-semibold'
         }
     ]
}|cva %}

<div {{ attributes.defaults({
    class: alert.apply({color, size}, class)
}) }}>
    ...
</div>

<div {{ alert.apply({color, size}, 'some other classes') }}>...</div>
```

smnandre commented:
Exactly !

kbond commented:
Best practices aside, this reads better to me:

```twig
{% set alert = cva({
    base: 'alert',
    variants: {
        color: {
            blue: 'alert-blue',
            red: 'alert-red',
            green: 'alert-green',
            yellow: 'alert-yellow',
        },
        size: {
            sm: 'alert-sm',
            md: 'alert-md',
            lg: 'alert-lg',
        }
    },
    compounds: [
        {
            color: ['red'],
            size: ['lg'],
            class: 'font-semibold'
        }
    ]
}) %}
```

Than:

```twig
{% set alert = {
    base: 'alert',
    variants: {
        color: {
            blue: 'alert-blue',
            red: 'alert-red',
            green: 'alert-green',
            yellow: 'alert-yellow',
        },
        size: {
            sm: 'alert-sm',
            md: 'alert-md',
            lg: 'alert-lg',
        }
    },
    compounds: [
        {
            color: ['red'],
            size: ['lg'],
            class: 'font-semibold'
        }
    ]
}|cva %}
```

But maybe that's just me.

Kocal commented:
Yes the first one is more easier to read and review

yguedidi commented:
if it get changed to a function, as it seems the first level of the structure is fixed, what about making them arguments?
```twig
{% set alertCva = cva(
    base='alert',
    variants={
        color: {
            blue: 'alert-blue',
            red: 'alert-red',
            green: 'alert-green',
            yellow: 'alert-yellow',
        },
        size: {
            sm: 'alert-sm',
            md: 'alert-md',
            lg: 'alert-lg',
        }
    },
    compounds=[
        {
            color: ['red'],
            size: ['lg'],
            class: 'font-semibold'
        }
    ]
) %}
```

so, without advanced compound it will look like this:
```twig
{% set alertCva = cva('alert', {
    color: {
        blue: 'alert-blue',
        red: 'alert-red',
        green: 'alert-green',
        yellow: 'alert-yellow',
    },
    size: {
        sm: 'alert-sm',
        md: 'alert-md',
        lg: 'alert-lg',
    }
}) %}
```

smnandre commented:
I understand your opinions on the syntax / readability. 

My point is: this would not be accepted as a function in twig/twig or twig/extra-**, probably neither in symfony/symfony.

We can 100% ignore that.. i'm just making sure it is an informed choice, especially if we later ask for Twig internal changes ;)

Kocal commented:
> So if it's a function it needs to be used in  `{{ cva(...) }}`
> But since it's only providing an object that needs to get applied() to actually create output, it should become a filter!

Why should it be a filter? Is this an opinionated rule or am I missing something...?

CMH-Benny commented:
> > So if it's a function it needs to be used in  `{{ cva(...) }}`
> > But since it's only providing an object that needs to get applied() to actually create output, it should become a filter!
> 
> Why should it be a filter? Is this an opinionated rule or am I missing something...?

As it was pointed out by @smnandre, it seems to be best practice. 

I myself, to be fully honest, don't really care so much about that. I also think the function reads/looks better, but if that breaks some conventions of twig then I am open to adapt to the filter. In the end I am just thankful that @WebMamba opened this PR, I am just looking forward to be able to use CVA in twig to build amazing components, so I am fine with either option, as long as it's merged at some point 😁

WebMamba commented:
> Since this isn't really twig-component specific (I don't think), should we at least try and get this into twig/html-extra? I know that is easier said than done.

@kbond this is an interesting proposal. I think to get the all potential of this PR is to use it in a laststack, so with TwigComponent installed. So I think in any case we should document this feature in the TwigComponent documentation. TwigComponent doesn't have a dependency on twig/html-extra do you think we should add it then? What do think about a new ux-extra package?

WebMamba commented:
@smnandre I see your point about the filter but here is not really `transforming` but more `creating` a CVA, so I think using function is still semantically correct.

WebMamba commented:
> if it get changed to a function, as it seems the first level of the structure is fixed, what about making them arguments?

Hummmm I am not sure about this. What do people think?

smnandre commented:
> `creating`

It's not about creating thing.. it's about generating output or not.. :)  Twig "things" that create new objects are filters (today).

Again, i just want you to take the decision with all information. :)

smnandre commented:
> TwigComponent doesn't have a dependency on twig/html-extra do you think we should add it then? What do think about a new ux-extra package?

I don't feel the absolute necessity TwigComponent is only about the "component" function (or HTML Syntax). 

If this `cva` filt.. function helps manage the style of _components_, it feel to me very legitimate to have it in the `ux-twig-component` package, no ?

smnandre commented:
> Why should it be a filter? Is this an opinionated rule or am I missing something...?

https://github.com/symfony/ux/pull/1416#issuecomment-1921941431

kbond commented:
> Hummmm I am not sure about this. What do people think?

I think this makes sense but the syntax for named parameters in twig is a bit unfortunate ([that might change in the future though](https://github.com/twigphp/Twig/issues/3635)).

kbond commented:
> TwigComponent doesn't have a dependency on twig/html-extra do you think we should add it then?

I'd actually be in favor of this as the `html_classes()` function from html-extra is quite useful.

> What do think about a new ux-extra package?

I don't think that's necessary. I'm totally fine having it live here for the time being. If it ever was added to html-extra, it'd likely be named `html_cva()` so there wouldn't be a conflict. At that point we could perhaps deprecate the function here.

Kocal commented:
> > Why should it be a filter? Is this an opinionated rule or am I missing something...?
> 
> [#1416 (comment)](https://github.com/symfony/ux/pull/1416#issuecomment-1921941431)

Thanks!

Well I think it depends of how you read it...? 🤔 

I'm more in favor on a function, it takes one (or many, depends of how you implement it) argument, a Twig object which defines base/variants, which returns a new `CVA` object

CMH-Benny commented:
So how to move forward now? To me it's kinda unclear, what needs to be done to proceed with this PR 😅 
Are there changes required? Is it fine and only needs the final review?

Would be awesome to have this in the next release, tho 🎉

yguedidi commented:
Just to share some thoughts about the filter vs function thing.

First, let's distinguish between "outputs" and "returns".
The only thing that really "outputs" is the use of `{{ }}`, it's doing the `echo`. We usually don't echo in functions nor filters, but we "return" a value.
Then, both of them deal with data, be it a string, a string-able object or any object.
A filter supports a single input. A function supports 0+ inputs (its arguments).
The forced single input in filters makes them good for "transforming" the data, they are like a funnel.
A function can be called with 0 or more arguments, and can "create" a data from nothing, or based on the arguments (and it could also transform one of its arguments like a filter).

To me, here we are "creating" a CVA object, based on some arguments, that's why I would say the function is a better fit here.
It's like the constructor of a CVA instance.
Being like a constructor to me, I'd not use an array to construct it (aaaah good-old `$options` arrays...) but precise arguments.

voila, hope this helps in the discussion/decision.

WebMamba commented:
I currently traveling for a couple of days but when I am back:
- I think to change to function argument, instead of having this array keys
- I created a PR on twig/html-extra in parallel with this one
- I think we can keep the function and not have a filter I agree with the argument brought by @yguedidi here.

Sounds good for everyone here?

smnandre commented on src/TwigComponent/src/CVA.php:
```diff
@@ -0,0 +1,99 @@
+    private ?string $base;
+
+    /**
+     * @var array<string, array<string, string>>
```
```suggestion
     * @var array<string, array<string, string>>|null
```

smnandre commented on src/TwigComponent/src/CVA.php:
```diff
@@ -0,0 +1,99 @@
+     */
+    private ?array $compounds;
+
+    public function __construct(string $base = null, array $variants = null, array $compounds = null)
```
Use property promotion ?

smnandre commented on src/TwigComponent/src/CVA.php:
```diff
@@ -0,0 +1,99 @@
+        $this->compounds = $compounds;
+    }
+
+    public static function init(array $recipe): self
```
Maybe a bit of PHP doc here to specify the `$recipe` format ?

smnandre commented on src/TwigComponent/src/CVA.php:
```diff
@@ -0,0 +1,99 @@
+        );
+    }
+
+    public function apply(array $recipes, string $classes): string
```
Some PHPDoc here ? 

Can string be empty ?

smnandre commented on src/TwigComponent/src/CVA.php:
```diff
@@ -0,0 +1,99 @@
+
+    public function apply(array $recipes, string $classes): string
+    {
+        return self::resolve($recipes).' '.$classes;
```
If resolve can render an empty string, or classes can be empty: 

```php
return trim(self::resolve($recipes).' '.$classes);
```

WDYT ?

smnandre commented on src/TwigComponent/src/CVA.php:
```diff
@@ -0,0 +1,99 @@
+
+    public function resolve(array $recipes): string
+    {
+        $classes = '';
+
+        if (null !== $this->base) {
+            $classes .= $this->base;
+        }
```
```php
$classes = $this->base ?? '';
```

wdyt ?

smnandre commented on src/TwigComponent/src/CVA.php:
```diff
@@ -0,0 +1,99 @@
+        }
+
+        foreach ($recipes as $recipeName => $recipeValue) {
+            if (!isset($this->variants[$recipeName])) {
+                continue;
+            }
+
+            if (!isset($this->variants[$recipeName][$recipeValue])) {
+                continue;
+            }
```
First test is useless

```
isset($this->variants[$recipeName][$recipeValue]
```

Will check for both at once

smnandre commented on src/TwigComponent/src/CVA.php:
```diff
@@ -0,0 +1,99 @@
+                }
+
+                if ($isCompound) {
+                    $classes .= ' '.$compound['class'];
```
Are we certain `compound[class]` is defined and is a stringable ?

smnandre commented on src/TwigComponent/src/Twig/ComponentExtension.php:
```diff
@@ -84,6 +86,11 @@ public function finishEmbeddedComponentRender(): void
         $this->container->get(ComponentRenderer::class)->finishEmbeddedComponentRender();
     }
 
+    public function cva(array $recipe): CVA
```
Some doctype there too if possible

smnandre commented on src/TwigComponent/tests/Fixtures/templates/components/Alert.html.twig:
```diff
@@ -0,0 +1,22 @@
+{% props color = 'blue', size = 'md', class = '' %}
+
+{{ cva('alert', {
```
This file is not used anymore in tests, right ?

smnandre commented on src/TwigComponent/tests/Unit/CVATest.php:
```diff
@@ -0,0 +1,215 @@
+
+        $this->assertEquals($expected, $recipeClass->resolve($recipes));
+    }
+
```
Could you maybe add one of two basic tests to check Exceptions / errors / invalid inputs ?

smnandre commented on src/TwigComponent/src/CVA.php:
```diff
@@ -0,0 +1,99 @@
+/**
+ * @author Mathéo Daninos <matheo.daninos@gmail.com>
+ */
+class CVA
```
Final and/or experimental here ? Or this will be under BC promise from day one :|

smnandre commented on src/TwigComponent/src/CVA.php:
```diff
@@ -0,0 +1,99 @@
+
+namespace Symfony\UX\TwigComponent;
+
+/**
```
Maybe add a short description of the concept and/or some reference to "standard"/"original" concepts (if any) ?

CMH-Benny commented:
Any updates on this? 😅

WebMamba commented:
Hey all! I just pushed https://github.com/symfony/ux/pull/1416#issuecomment-1936354904 the following change, add a line to the change log and some documentation. Cheers 🧡

weaverryan commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,143 @@ Exclude specific attributes:
       My Component!
     </div>
 
+.. __Customize your component:
```
I don't think (?) this needs to be there

weaverryan commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,143 @@ Exclude specific attributes:
+CVA (Class variant authority)
+-----------------------------
+
+You can use the cva function to define variants for your component.
```
Class Variant Authority needs a quick explanation and maybe a link. Also if you don't know what CVA is (few will) the header isn't super helpful. Maybe this?

```
Components with Complex Variants (CVA)
```

weaverryan commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,143 @@ Exclude specific attributes:
+    {% props color = 'blue', size = 'md', class = '' %}
+
+     {% set alert = cva('alert', {
+        'color': {
```
Remove the `'` on all of the array keys

weaverryan commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,143 @@ Exclude specific attributes:
+.. code-block:: html+twig
+
+    {# templates/components/Alert.html.twig #}
+    {% props color = 'blue', size = 'md', class = '' %}
```
What's the point of the `class` prop? Is that necessary or can we just let the normal attributes system handle things if `class` is passed in?

CMH-Benny replied:
I think this is still coming from earler, when there was also a tailwind merge baked in, one could just leave it out and use it as normal attribute. 

But maybe someone puts custom tailwind in "class" attribute of the component and expects they are merged with the ones the component sets based on variants, I guess if that is the goal it's required to make class a prop and then pass it as second argument to apply and then use tw_merge from other bundle to clean up classes before actually rendering it in the class attribute of the element inside of the component :) 

Maybe docs could show both as an example?

weaverryan commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,143 @@ Exclude specific attributes:
+    }) %}
+
+    <div {{ attributes.defaults(
+        class: alert.apply({color, size}, class)
```
how do "base classes" work - i.e. classes that will ALWAYS be needed on this component? Would I do something like this?

```twig
class: alert.apply({color, size}, 'the-base-class another-base-class')
```

Or is there a `base` key that's not in the docs yet?

WebMamba replied:
The base class is simply a set of classes present in all variations of your components: for example, for a button, you may want all your variants to contain the class 'btn'. It's the first argument of the cva function. I added some doc for it! Thanks!

CMH-Benny replied:
cva twig functions first parameter are the `base` classes, they are always applied and get combined with the variants - I think having that option is great. It could be done as in your example as well, but what if you also want to ask for additional classes from "outside" (via the `<twig:X class="...">`? 
It could be concatenated with the 2nd parameter of apply() for sure, but the second param of apply is ment as "additional classes"

Base classes are part of CVA and even tho it would be solvable without `base` classes, shouldn't we stick to how CVA works in JS? https://cva.style/docs/api-reference

`defaultVariants` were already left out, because those can be handled by default values in props that are passed as variants, but do we really also want to scratch `base` classes?

weaverryan commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,143 @@ Exclude specific attributes:
+    }) %}
+
+    <div {{ attributes.defaults(
+        class: alert.apply({color, size}, class)|tailwind_merge
```
Would this example (and maybe CVA in general) look & work better using Kevin's attribute rendering system? https://github.com/symfony/ux/pull/1442

weaverryan commented on src/TwigComponent/src/CVA.php:
```diff
@@ -0,0 +1,88 @@
+/**
+ * @author Mathéo Daninos <matheo.daninos@gmail.com>
+ *
+ * CVA (class variant authority), is a concept highly use in the js word.
```
```suggestion
 * CVA (class variant authority), is a concept from the js world.
```

weaverryan commented on src/TwigComponent/src/CVA.php:
```diff
@@ -0,0 +1,88 @@
+ *
+ * CVA (class variant authority), is a concept highly use in the js word.
+ * https://cva.style/docs
+ * The UI library shadcn is build on top of this principles
```
```suggestion
 * The UI library shadcn is build on top of this principle
```

weaverryan commented on src/TwigComponent/src/CVA.php:
```diff
@@ -0,0 +1,88 @@
+ * https://cva.style/docs
+ * The UI library shadcn is build on top of this principles
+ * https://ui.shadcn.com
+ * The concept behind CVA is to let you build component with a lot of different variation call recipes.
```
```suggestion
 * The concept behind CVA is to let you build component with a lot of different variations called recipes.
```

weaverryan commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,143 @@ Exclude specific attributes:
+CVA and Tailwind
+~~~~~~~~~~~~~~~~
+
+CVA work perfectly with tailwindcss. The only drawback is you can have class conflicts,
```
Tailwind CSS

weaverryan replied:
Still need to fixing the casing of Tailwind in this sentence :)

weaverryan replied:
```suggestion
CVA work perfectly with tailwindcss. The only drawback is that you can have class conflicts.
```

weaverryan commented on src/TwigComponent/src/CVA.php:
```diff
@@ -0,0 +1,88 @@
+
+    public function apply(array $recipes, ?string $classes = null): string
+    {
+        return trim(self::resolve($recipes).' '.$classes ?? '');
```
`$this->resolve(`

weaverryan commented on src/TwigComponent/src/CVA.php:
```diff
@@ -0,0 +1,88 @@
+
+                if ($isCompound) {
+                    if (!isset($compound['class']) || !\is_string($compound['class'])) {
+                        throw new \LogicException('compound recipe are matching but no class are register for this match');
```
```suggestion
                        throw new \LogicException('A compound recipe matched but no classes are registered for this match');
```

weaverryan reviewed (changes requested):
I like this. Let's smooth some edges and get it done 👍

Kocal commented on src/TwigComponent/tests/Fixtures/templates/components/Alert.html.twig:
```diff
@@ -0,0 +1,26 @@
+{% props color = 'blue', size = 'md' %}
+
+{% set alert = cva('alert', {
```
So the first argument is still a thing? What does it do?

WebMamba replied:
https://github.com/symfony/ux/pull/1416#discussion_r1498352255
Add some bases classes 😁

Kocal replied:
Oh alright, I didn't see it thanks. 

Do you think it really should be mandatory?

WebMamba replied:
It's not mandatory you can have an empty value, it's just we don't have named arguments in Twig, but none of those arguments should be mandatory

WebMamba commented:
@weaverryan thanks for your review 🧡 I addressed your comment! Sorry about the syntax mistake, I make this kind of error even in French 😭 
@kbond  do I use the arguments functions the right way? Is it what you have in mind?

kbond commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,141 @@ Exclude specific attributes:
+Component with Complex Variants (CVA)
+-------------------------------------
+
+CVA is a concept from the JS words (https://cva.style/docs/getting-started/variants).
```
```suggestion
CVA is a concept from the JS world (https://cva.style/docs/getting-started/variants).
```

?

kbond commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,141 @@ Exclude specific attributes:
+-------------------------------------
+
+CVA is a concept from the JS words (https://cva.style/docs/getting-started/variants).
+It's a concepts use by the famous shadcn/ui library (https://ui.shadcn.com).
```
```suggestion
It's a conceptd used by the famous shadcn/ui library (https://ui.shadcn.com).
```

smnandre replied:
conceptd ?

kbond commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,141 @@ Exclude specific attributes:
+
+CVA is a concept from the JS words (https://cva.style/docs/getting-started/variants).
+It's a concepts use by the famous shadcn/ui library (https://ui.shadcn.com).
+This concept allow you to display a component with different variants (color, size, etc.),
```
```suggestion
This concept allows you to display a component with different variants (color, size, etc.),
```

kbond commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,141 @@ Exclude specific attributes:
+        }
+    }) %}
+
+    <div class="{{ alert.apply({color, size}) ~ ' ' ~ attributes.render('class') | tailwind_merge }}">
```
Does `tailwind_merge` apply to `attributes.render('class')` only? Either way, it's a bit confusing imo.

What about rendering the class above?

```diff
- {% set alert = cva('alert', {
+ {% set alert = cva('alert ' ~ attributes.render('class'), {
```

Then, here:

```suggestion
    <div class="{{ alert.apply({color, size})|tailwind_merge }}">
```

WebMamba replied:
Yes good idea, it's cleaner! 👍

CMH-Benny replied:
But in this example you can't override anything that is set via variants from outside anymore, right? I mean maybe that is intended, but if you want to allow to override the final look of a component, the outer classes needs to go as 2nd argument to apply, correct?

kbond replied:
Can you explain this a bit more? I don't see why you couldn't still set variants from outside

WebMamba replied:
@CMH-Benny I don't think it's changing something. The only difference is now the outside class gonna be in the center of the other class and before the outside class was at the end.

CMH-Benny replied:
Okay, let me try it with an example and how i understand it, I mean in the end this is not a blocker, I just wasn't sure if we should mention the behavior so that devs can do it the way they need it :) 

Let's say we have component:
```twig
{% props theme='primary', size = 'md' %}
{% set button = cva('btn' ~ ' ' ~ attributes.render('class'), {
         theme: {
             primary: 'bg-blue-500',
             secondary: 'bg-gray-400',
             success: 'bg-green-300',
             warning: 'bg-yellow-500',
             danger: 'bg-red-600',
         },
         size: {
             sm: 'p-2',
             md: 'p-4',
             lg: 'p-8',
         }
     },
     compounds: [
         {
             theme: ['danger'],
             size: ['lg'],
             class: 'font-semibold'
         }
     ]
}) %}

<button {{ button.apply({theme, size}, 'some other classes')|tailwind_merge }}>...</button>
```

Now using it somewhere:
```twig
<twig:Button theme="danger" size="sm" class="bg-[#3d3d3d]" />
```

Now, background of that button will still be red and not the custom one - Can be intended, but could also be not intuitive, but since attributes appends classes to base classes, the variant class comes later and probably wins in tailwind_merge over the custom color, or am I wrong here?

kbond replied:
Ah yes, thanks for the explanation. This is a indeed a problem when using tailwind_merge.

weaverryan replied:
Would this work?

```twig
{% props theme='primary', size = 'md' %}
{% set button = cva('btn', {
         theme: {
             primary: 'bg-blue-500',
             secondary: 'bg-gray-400',
             success: 'bg-green-300',
             warning: 'bg-yellow-500',
             danger: 'bg-red-600',
         },
         size: {
             sm: 'p-2',
             md: 'p-4',
             lg: 'p-8',
         }
     },
     compounds: [
         {
             theme: ['danger'],
             size: ['lg'],
             class: 'font-semibold'
         }
     ]
}) %}

<button class="{{ button.apply({theme, size}, 'some other classes', attributes.render('class'))|tailwind_merge }}">...</button>
```

The `apply()` method would change to allow for an unlimited number of "class" arguments. It's a bit verbose is my only complaint (assuming that this also would work).

I suppose another option could be to allow the `CVA` object be mutated via a `.classes` method at the end

```twig
{% props theme='primary', size = 'md' %}
{% set button = cva('btn', {
     }
    ....
).classes('some other classes', attributes.render('class'))
}) %}
```

kbond replied:
> The apply() method would change to allow for an unlimited number of "class" arguments.

I like this!

CMH-Benny commented:
Amazing progress, I am looking forward to be able to use this feature!
Thank you so much everyone! 😄

kbond commented:
@WebMamba, is the PR description in sync with the current code?

> `{% set alert =  cva('alert', 'rounded-lg',`

The second arg cannot be a string, correct?

kbond commented:
For this example, is both `alert` and `rounded-lg` considered the "base"?

```twig
{% set alert = cva('alert', {
    base: 'rounded-lg',
    variants: {
        color: {
            blue: 'text-blue-800 bg-blue-50 dark:bg-gray-800 dark:text-blue-400',
            red: 'text-red-800 bg-red-50 dark:bg-gray-800 dark:text-red-400',
            green: 'text-green-800 bg-green-50 dark:bg-gray-800 dark:text-green-400',
            yellow: 'text-yellow-800 bg-yellow-50 dark:bg-gray-800 dark:text-yellow-400',
        },
        size: {
            sm: 'px-4 py-3 text-sm',
            md: 'px-6 py-4 text-base',
            lg: 'px-8 py-5 text-lg',
        }
    }
}) %}
```

What about making the only argument to the cva twig function an array so we can use it as quasi-named parameters:

```twig
{% set alert =  cva({
    base: 'alert rounded-lg', {# or ['alert', 'rounded-lg'] #}
    variants: {
        color: {
            blue: 'text-blue-800 bg-blue-50 dark:bg-gray-800 dark:text-blue-400',
            red: 'text-red-800 bg-red-50 dark:bg-gray-800 dark:text-red-400',
            green: 'text-green-800 bg-green-50 dark:bg-gray-800 dark:text-green-400',
            yellow: 'text-yellow-800 bg-yellow-50 dark:bg-gray-800 dark:text-yellow-400',
        },
        size: {
            sm: 'px-4 py-3 text-sm',
            md: 'px-6 py-4 text-base',
            lg: 'px-8 py-5 text-lg',
        }
    }
}) %}
```

Is this intended to be a port of https://cva.style/docs/api-reference? If so, you use the term _compound_ but cva.style has _compoundVariants_ - should we match? Also, cva.style has _defaultVariants_?

WebMamba commented:
@kbond 
> @WebMamba, is the PR description in sync with the current code?

Big sorry I made a typo yesterday while editing the description 😬

> What about making the only argument to the cva twig function an array so we can use it as quasi-named parameters

This was my original proposal but we asked me to use named arguments instead, and I think it would be a good idea if Twig has named argument as you mention here: https://github.com/symfony/ux/pull/1416#issuecomment-1925417508

> Is this intended to be a port of https://cva.style/docs/api-reference? If so, you use the term compound but cva.style has compoundVariants - should we match? Also, cva.style has defaultVariants?

Thanks, I gonna make some changes to match the original name

@weaverryan 
> The apply() method would change to allow for an unlimited number of "class" arguments. It's a bit verbose is my only complaint (assuming that this also would work).

I like this idea thanks! I much prefer this approach to the .class. So let's try it!

kbond commented:
> Thanks, I gonna make some changes to match the original name

If you use these _camelCase_ names (`compoundVariants` & `defaultVariants`), using twig named parameters is going to suffer from https://github.com/twigphp/Twig/issues/3475. Maybe this would just be a documentation thing but my preference would be to use an array only. If in the future, named arguments are fixed/improved in twig, we should be able to switch w/o BC break.

WebMamba commented:
OK, thanks @kbond! I am on it

kbond commented on src/TwigComponent/src/Twig/ComponentExtension.php:
```diff
@@ -84,6 +86,29 @@ public function finishEmbeddedComponentRender(): void
+    public function cva(array $cva): CVA
+    {
+        return new CVA(
+            $cva['base'] ?? [],
```
```suggestion
            $cva['base'] ?? null,
```

kbond commented on src/TwigComponent/src/Twig/ComponentExtension.php:
```diff
@@ -84,6 +86,29 @@ public function finishEmbeddedComponentRender(): void
 
+    /**
+     * @param array{
+     *     base: string,
```
```suggestion
     *     base: string|string[]|null,
```

kbond commented on src/TwigComponent/src/CVA.php:
```diff
@@ -0,0 +1,97 @@
+     * @var array<array<string, string[]>>|null       the array should have the following format ['variantsCategory' => ['variantName', 'variantName'], 'class' => 'text-red-500']
+     */
+    public function __construct(
+        private ?string $base = null,
```
```suggestion
        private string|array|null $base = null,
```

The base being a `list<string|null>` needs to be accounted in the logic below.

kbond commented on src/TwigComponent/src/CVA.php:
```diff
@@ -0,0 +1,97 @@
+final class CVA
+{
+    /**
+     * @var string|null
```
```suggestion
     * @var string|list<string|null>|null
```

BTW, I'm suggesting this to help with https://github.com/symfony/ux/pull/1416#discussion_r1506202909

kbond commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,141 @@ Exclude specific attributes:
+    {# templates/components/Alert.html.twig #}
+    {% props color = 'blue', size = 'md' %}
+
+     {% set alert = cva('alert ' ~ attributes.render('class'), {
+        color: {
+            blue: 'bg-blue',
+            red: 'bg-red',
+            green: 'bg-green',
+        },
+        size: {
+            sm: 'text-sm',
+            md: 'text-md',
+            lg: 'text-lg',
+        }
+    }) %}
```
```suggestion
     {% set alert = cva({
        base: ['alert', attributes.render('class')], {
        color: {
            blue: 'bg-blue',
            red: 'bg-red',
            green: 'bg-green',
        },
        size: {
            sm: 'text-sm',
            md: 'text-md',
            lg: 'text-lg',
        }
    }) %}
```

WebMamba replied:
I made the change so the base now accepts array. But I think is easier if you just use the function apply to append your class.
`cva.apply(cva, attributes.render('class'))`

kbond replied:
Yes, I agree, but would base be the place to prepend classes if that's your intention?

WebMamba replied:
Yes this why I added the ability to have list for base, I think your right it can be useful in some situation

kbond replied:
But anyway, this matches [cva.style](https://cva.style/docs/api-reference#parameters)

kbond commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,141 @@ Exclude specific attributes:
+CVA allows you to display a component with different variants (color, size, etc.),
+to create highly reusable and customizable components.
+You can use the cva function to define variants for your component.
+The first argument of the cva function is the base classes, the classes present in all variants. Then the second
```
This verbiage will need to be updated.

kbond commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,141 @@ Exclude specific attributes:
+    {# templates/components/Alert.html.twig #}
+    {% props color = 'blue', size = 'md' %}
+
+     {% set alert = cva('alert ' ~ attributes.render('class'), {
+        color: {
+            blue: 'bg-blue',
+            red: 'bg-red',
+            green: 'bg-green',
+        },
+        size: {
+            sm: 'text-sm',
+            md: 'text-md',
+            lg: 'text-lg',
+        }
+    }) %}
```
```suggestion
     {% set alert = cva({
        base: 'alert',
        color: {
            blue: 'bg-blue',
            red: 'bg-red',
            green: 'bg-green',
        },
        size: {
            sm: 'text-sm',
            md: 'text-md',
            lg: 'text-lg',
        }
    }) %}
```

kbond commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,141 @@ Exclude specific attributes:
+        }
+    }) %}
+
+    <div class="{{ alert.apply({color, size}) | tailwind_merge }}">
```
```suggestion
    <div class="{{ alert.apply([color, size], attributes.render('class')) | tailwind_merge }}">
```

1. Should the first argument for apply be a list and not a hash?
2. This ensures the passed classes are always last - this is the common behaviour, correct?

Examples below should be updated similarly.

WebMamba replied:
It can't be a list because we use the key to find the variant. Do you see what I mean?

kbond replied:
Hmm, what does `{color, size}` convert to in php? `['color' => null, 'size' => null]`?

WebMamba replied:
Yes exactly. I didn't know that, but Ryan showed me the trick

kbond replied:
Oh wait, `{color, size}` actually converts to `['color' => 'blue', 'size' => 'md']` in the context of the example in the PR description. Am I correct here?

CMH-Benny replied:
Yes, `{color, size}` it basically converts to:

```
{
 color: color,
 size: size
}
```
Similar to JS as far as I understood it, so in the end via resolved values it will turn into
```
{
 'color': 'blue',
 'size': 'md'
}
```

Which becomes `['color' => 'blue', 'size' => 'md']` in PHP

kbond reviewed (changes requested):
Getting close I think - just a few suggestions.

WebMamba commented:
Hey all! I updated the description of the PR but here is what the last commit brings:
So now the CVA function has just one argument a big array key:
- base: to define all the classes that are present in all variations, it's a string
- variants: your different variants
- compoundVariants: let you the ability to apply a set of classes only if a specific set of variants match
- defaultVariants: let you define a default variant to pick if none of the variations match

The function `apply can accept has many arguments you want to give you the ability to add as many sets of class as you want.
And I made some renaming.
Cheers 🧡

kbond reviewed (approved):
🚀

weaverryan commented on src/TwigComponent/CHANGELOG.md:
```diff
@@ -5,12 +5,14 @@
 -   Add the ability to render specific attributes from the `attributes` variable #1442
 -   Restrict Twig 3.9 for now #1486
 -   Build reproducible TemplateMap to fix possible post-deploy breakage #1497
+-   Add CVA (Class variant authority) integration #1416
```
Should be on 2.16.0

weaverryan commented on src/TwigComponent/CHANGELOG.md:
```diff
@@ -5,12 +5,14 @@
 -   Make `ComponentAttributes` traversable/countable
 -   Fixed lexing some `{# twig comments #}` with HTML Twig syntax
 -   Fix various usages of deprecated Twig code
+-   Add attribute rendering system
```
This slipped in - should not be here

weaverryan commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,198 @@ Exclude specific attributes:
+Component with Complex Variants (CVA)
+-------------------------------------
+
+CVA (Class Variant Authority) is a concept from the JS world (https://cva.style/docs/getting-started/variants).
```
```suggestion
[CVA (Class Variant Authority)](https://cva.style/docs/getting-started/variants) is a concept from the JavaScript
world and used by the well-known [shadcn/ui](https://ui.shadcn.com).
```

weaverryan commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,198 @@ Exclude specific attributes:
+It's a concept used by the famous shadcn/ui library (https://ui.shadcn.com).
+CVA allows you to display a component with different variants (color, size, etc.),
+to create highly reusable and customizable components.
+You can use the cva function to define variants for your component.
```
```suggestion
This is powered by a ``cva()`` Twig function where you define ``base`` classes that should always be present and then different ``variants`` and the corresponding classes:
```

weaverryan commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,198 @@ Exclude specific attributes:
+    <div class="{{ alert.apply({color, size}, attributes.render('class')) }}">
+         {% block content %}{% endblock %}
+    </div>
+
```
Break the code blocks into 2 with some text:

Then use the ``color`` and ``size`` variants to select the classes needed:

weaverryan commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,198 @@ Exclude specific attributes:
+    </twig:Alert>
+    // class="alert bg-green text-sm"
+
+    <twig:Alert class="flex items-center justify-center">
```
Show a `color` attribute here so we show that you can pass the variants AND custom classes..

weaverryan commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,198 @@ Exclude specific attributes:
+~~~~~~~~~~~~~~~~~~~~
+
+CVA work perfectly with tailwindcss. The only drawback is you can have class conflicts,
+to have a better control you can use this following bundle (
```
```suggestion
To "merge" conflicting classes together and keep only the one you need, use the
``tailwind_merge()` method from [tales-from-a-dev/twig-tailwind-extra](https://github.com/tales-from-a-dev/twig-tailwind-extra)
with the ``cva()`` function:
```

weaverryan commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,198 @@ Exclude specific attributes:
+    {% props color = 'blue', size = 'md' %}
+
+   {% set alert = cva({
+        base: 'alert ',
```
I think we can just put `// ...` inside the `cva()` function instead of listing `base` and `variants` again.

weaverryan commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,198 @@ Exclude specific attributes:
+         {% block content %}{% endblock %}
+    </div>
+
+Compounds variants
```
```suggestion
Compounds Variants
```

CMH-Benny replied:
*`Compound Variants` or `Compounded Variants` is probably even better?

weaverryan commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,198 @@ Exclude specific attributes:
+Compounds variants
+~~~~~~~~~~~~~~~~~~
+
+You can define compound variants. A compound variant is a variants that apply
```
```suggestion
A compound variant is a variant that applies
```

weaverryan commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,198 @@ Exclude specific attributes:
+            }
+        },
+        compound: {
+            colors: ['red'],
```
Maybe a comment above this would help describe it - e.g.

```
// if colors=red AND size = (md or lg), add the `font-bold` class
```

weaverryan commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,198 @@ Exclude specific attributes:
+    </twig:Alert>
+    // class="alert bg-green text-lg font-bold"
+
+Default variants
```
```suggestion
Default Variants
```

weaverryan commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,198 @@ Exclude specific attributes:
+Default variants
+~~~~~~~~~~~~~~~~
+
+You can define defaults variants, so if no variants are matching you
```
```suggestion
If no variants match, you can define a default set of classes to apply:
```

weaverryan commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,198 @@ Exclude specific attributes:
+.. code-block:: html+twig
+
+    {# templates/components/Alert.html.twig #}
+    {% props color = 'blue', size = 'md' %}
```
Do we need `rounded = ''` up here?


And if so, then does `defaultVariants` have limited usage? Not that we should remove it, just clarifying :)

WebMamba replied:
CVA can be use outside of a component this is why we may need it. What do you think ?

weaverryan commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,198 @@ Exclude specific attributes:
+                lg: 'rounded-lg',
+            }
+        },
+        defaultsVariants: {
```
```suggestion
        defaultVariants: {
```

weaverryan commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1058,6 +1058,198 @@ Exclude specific attributes:
     </div>
 
+Component with Complex Variants (CVA)
+-------------------------------------
```
Need a `versionadded` for 2.16

weaverryan commented on src/TwigComponent/tests/Unit/CVATest.php:
```diff
@@ -0,0 +1,339 @@
+    {
+        $recipeClass = new CVA($recipe['base'] ?? '', $recipe['variants'] ?? [], $recipe['compounds'] ?? [], $recipe['defaultVariants'] ?? []);
+
+        $this->assertEquals($expected, $recipeClass->resolve($recipes));
```
Missing a basic test with `->apply()` that shows the classes being appending properly and trimmed. It's tested via the Twig function test, but better to have a simple one here too.

weaverryan commented:
Thank you for this huge work Matheo! I've left some minor comments and docs comments for a future PR please :). But let's get this merged and in there

CMH-Benny commented:
Wooohoo, it's merged 🎉 

Thank you everyone for your efforts and the amazing features you provide to us developers ❤️

WebMamba commented:
Thanks a lot, everyone for participating in this PR! I wish you a lot of fun with this cool new feature! 🧡

Kocal commented:
Great! :D 

Thanks everyooone :)

seb-jean commented:
@WebMamba  How could I handle booleans?

Here is an example but it doesn't work

```twig
{# templates/components/Avatar.html.twig #}

{% props square = false, src, alt = null %}

 {% set avatar = cva({
     base: 'inline-grid align-middle *:col-start-1 *:row-start-1 ',
     variants: {
         square: {
             true: 'rounded-[20%] *:rounded-[20%]',
             false: 'rounded-full *:rounded-full',
         },
     }
 }) %}
```

WebMamba commented:
Hey @seb-jean, the following code should work: 
`avartar.apply({square: square ? 'true' : 'false'})`

seb-jean commented:
Yeah !! Thanks @WebMamba 🏀 !

seb-jean commented:
But is this the best way to handle boolean variants?
